tests/devapps/MSI/HelperService/MSIHelperService/Helper/MSIHelper.cs

```diff
@@ -0,0 +1,370 @@
+        /// </summary>
+        /// <param name="logger"></param>
+        /// <returns></returns>
+        public static Dictionary<string, string>? GetFunctionAppEnvironmentVariables(
+            ILogger logger)
+        {
+            logger.LogInformation("GetFunctionAppEnvironmentVariables Function called.");
+
+            string? token = Task.Run(async () => await GetMSALToken(logger).ConfigureAwait(false)).GetAwaiter().GetResult();
+
+            //clear the default request header for each call
+            ClearDefaultRequestHeaders(logger);
+
+            //Set the Authorization header
+            SetAuthorizationHeader(token, logger);
+
+            //send the request
+            HttpResponseMessage result = Task.Run(async () => await s_httpClient
+            .GetAsync(s_functionAppUri + "GetEnvironmentVariables?code=" + s_functionAppEnvCode).ConfigureAwait(false))
+            .GetAwaiter().GetResult();
+
+            var content = Task.Run(
+                async () => await result.Content.ReadAsStringAsync().ConfigureAwait(false))
+                .GetAwaiter().GetResult();
+
+            Dictionary<string, string>? envValuePairs = JsonConvert.DeserializeObject<Dictionary<string, string>>(content);
+
+            logger.LogInformation("GetFunctionAppEnvironmentVariables call was successful.");
+
+            return envValuePairs;
+        }
+
+        /// <summary>
+        /// Gets the MSI Token from the Azure Web App
+        /// </summary>
+        /// <param name="identityHeader"></param>
+        /// <param name="uri"></param>
+        /// <param name="logger"></param>
+        /// <returns></returns>
+        public static async Task<ActionResult?> GetWebAppMSIToken(
+            string? identityHeader,
+            string? uri,
+            ILogger logger)
+        {
+            logger.LogInformation("GetWebAppMSIToken Function called.");
+
+            var decodedUri = HttpUtility.UrlDecode(uri);
+
+            //set the http get method and the required headers for a web app
+            var requestMessage = new HttpRequestMessage(HttpMethod.Get, decodedUri);
+            requestMessage.Headers.Add(ManagedIdentityAuthenticationHeader, identityHeader);
+
+            //send the request
+            HttpResponseMessage result = await s_httpClient.SendAsync(requestMessage).ConfigureAwait(false);
+
+            string body = await result.Content.ReadAsStringAsync().ConfigureAwait(false);
+
+            logger.LogInformation("GetWebAppMSIToken Function call was successful.");
+
+            return GetContentResult(body, "application/json", (int)result.StatusCode);
+        }
+
+        /// <summary>
+        /// Gets the MSI Token from the Azure Function App
+        /// </summary>
+        /// <param name="identityHeader"></param>
+        /// <param name="uri"></param>
+        /// <param name="logger"></param>
+        /// <returns></returns>
+        public static async Task<ActionResult?> GetFunctionAppMSIToken(
```

Comment:
How is this going to play with MSAL? Because MSAL allows injection of HttpClient, but you are returning a full response here (ManagedIdentityResponse)

Reply:
Added the integration test PR, please take a look at how we consume this MSIHelper service within MSAL